libfreerdp/core/childsession.c

```diff
@@ -0,0 +1,332 @@
+}
+
+static int transport_bio_named_uninit(BIO* bio)
+{
+	WINPR_BIO_NAMED* ptr = (WINPR_BIO_NAMED*)BIO_get_data(bio);
+
+	if (ptr && ptr->hFile)
+	{
+		CloseHandle(ptr->hFile);
+		ptr->hFile = NULL;
+	}
+
+	BIO_set_init(bio, 0);
+	BIO_set_flags(bio, 0);
+	return 1;
+}
+
+static int transport_bio_named_new(BIO* bio)
+{
+	BIO_set_flags(bio, BIO_FLAGS_SHOULD_RETRY);
+
+	WINPR_BIO_NAMED* ptr = (WINPR_BIO_NAMED*)calloc(1, sizeof(WINPR_BIO_NAMED));
+	if (!ptr)
+		return 0;
+
+	BIO_set_data(bio, ptr);
+	BIO_set_flags(bio, BIO_FLAGS_SHOULD_RETRY);
+	return 1;
+}
+
+static int transport_bio_named_free(BIO* bio)
+{
+	WINPR_BIO_NAMED* ptr;
+
+	if (!bio)
+		return 0;
+
+	transport_bio_named_uninit(bio);
+	ptr = (WINPR_BIO_NAMED*)BIO_get_data(bio);
+
+	if (ptr)
+	{
+		BIO_set_data(bio, NULL);
+		free(ptr);
+	}
+
+	return 1;
+}
+
+BIO_METHOD* BIO_s_namedpipe(void)
+{
+	static BIO_METHOD* bio_methods = NULL;
+
+	if (bio_methods == NULL)
+	{
+		if (!(bio_methods = BIO_meth_new(BIO_TYPE_NAMEDPIPE, "NamedPipe")))
+			return NULL;
+
+		BIO_meth_set_write(bio_methods, transport_bio_named_write);
+		BIO_meth_set_read(bio_methods, transport_bio_named_read);
+		BIO_meth_set_puts(bio_methods, transport_bio_named_puts);
+		BIO_meth_set_gets(bio_methods, transport_bio_named_gets);
+		BIO_meth_set_ctrl(bio_methods, transport_bio_named_ctrl);
+		BIO_meth_set_create(bio_methods, transport_bio_named_new);
+		BIO_meth_set_destroy(bio_methods, transport_bio_named_free);
+	}
+
+	return bio_methods;
+}
+
+typedef NTSTATUS (*WinStationCreateChildSessionTransportFn)(WCHAR* path, DWORD len);
+BOOL createChildSessionTransport(HANDLE* pFile)
+{
+	WINPR_ASSERT(pFile);
+
+	BOOL ret = FALSE;
+	*pFile = INVALID_HANDLE_VALUE;
+
+	BOOL childEnabled;
+	if (!WTSIsChildSessionsEnabled(&childEnabled))
+	{
+		WLog_ERR(TAG, "error when calling WTSIsChildSessionsEnabled");
+		goto out;
+	}
+
+	if (!childEnabled)
+	{
+		WLog_INFO(TAG, "child sessions aren't enabled");
+		if (!WTSEnableChildSessions(TRUE))
+		{
+			WLog_ERR(TAG, "error when calling WTSEnableChildSessions");
+			goto out;
+		}
+		WLog_INFO(TAG, "successfully enabled child sessions");
+	}
+
+	HANDLE hModule = LoadLibraryA("winsta.dll");
+	if (!hModule)
+		return FALSE;
+	WCHAR pipePath[0x80];
+	char pipePathA[0x80];
+
+	WinStationCreateChildSessionTransportFn createChildSessionFn =
+	    (WinStationCreateChildSessionTransportFn)GetProcAddress(
+	        hModule, "WinStationCreateChildSessionTransport");
+	if (!createChildSessionFn)
+	{
+		WLog_ERR(TAG, "unable to retrieve WinStationCreateChildSessionTransport function");
+		goto out;
+	}
+
+	HRESULT hStatus = createChildSessionFn(pipePath, 0x80);
+	if (!SUCCEEDED(hStatus))
+	{
+		WLog_ERR(TAG, "error 0x%x when creating childSessionTransport", hStatus);
+		goto out;
+	}
+
+	ConvertWCharToUtf8(pipePath, pipePathA, sizeof(pipePathA));
+	WLog_DBG(TAG, "child session is at '%s'", pipePathA);
+
+	HANDLE f = CreateFileW(pipePath, GENERIC_READ | GENERIC_WRITE, 0, (LPSECURITY_ATTRIBUTES)0x0,
```

Comment:
useless casts, but use `NULL` instead of `0x0`

Reply:
Also fixed the NULL LPSECURITY_ATTRIBUTES